File: backend/app/schemas/rag.py
from pydantic import BaseModel, Field


class RagQueryRequest(BaseModel):
    question: str = Field(min_length=1, max_length=2000)
    # Optionally restrict retrieval to specific uploaded documents. When omitted,
    # all of the current user's processed documents are searched.
    document_ids: list[int] | None = None
    top_k: int | None = Field(default=None, ge=1, le=10)


class RagSource(BaseModel):
    document_id: int
    chunk_index: int
    snippet: str


class RagAnswerResponse(BaseModel):
    answer: str
    sources: list[RagSource]

File: frontend/flutter_app/lib/models/rag.dart
class RagSource {
  final int documentId;
  final int chunkIndex;
  final String snippet;

  const RagSource({
    required this.documentId,
    required this.chunkIndex,
    required this.snippet,
  });

  factory RagSource.fromJson(Map<String, dynamic> json) {
    return RagSource(
      documentId: json['document_id'] as int,
      chunkIndex: json['chunk_index'] as int,
      snippet: json['snippet'] as String? ?? '',
    );
  }
}

class RagAnswer {
  final String answer;
  final List<RagSource> sources;

  const RagAnswer({required this.answer, required this.sources});

  factory RagAnswer.fromJson(Map<String, dynamic> json) {
    final rawSources = json['sources'] as List<dynamic>? ?? [];
    return RagAnswer(
      answer: json['answer'] as String? ?? '',
      sources: rawSources
          .map((s) => RagSource.fromJson(s as Map<String, dynamic>))
          .toList(),
    );
  }
}
